src/modules/physics/physics_jolt.c

```diff
@@ -0,0 +1,1474 @@
+  return world->collision_steps;
+}
+
+void lovrWorldSetStepCount(World* world, int iterations) {
+  // todo: with too big count JobSystemThreadPool.cpp:124: (false) No jobs available!
+  world->collision_steps = iterations;
+}
+
+void lovrWorldComputeOverlaps(World* world) {
+  //arr_clear(&world->overlaps);
+}
+
+int lovrWorldGetNextOverlap(World* world, Shape** a, Shape** b) {}
+
+int lovrWorldCollide(World* world, Shape* a, Shape* b, float friction, float restitution) {}
+
+void lovrWorldGetContacts(World* world, Shape* a, Shape* b, Contact contacts[MAX_CONTACTS], uint32_t* count) {}
+
+void lovrWorldRaycast(World* world, float x1, float y1, float z1, float x2, float y2, float z2, RaycastCallback callback, void* userdata) {
+  const JPH_NarrowPhaseQuery* query = JPC_PhysicsSystem_GetNarrowPhaseQueryNoLock(world->physics_system);
+  const JPH_RVec3 origin = {.x = x1, .y = y1, .z = z1};
+  const JPH_Vec3 direction = {.x = x2 - x1, .y = y2 - y1, .z = z2 - z1};
+  JPH_AllHit_CastRayCollector* collector = JPH_AllHit_CastRayCollector_Create();
+  JPH_NarrowPhaseQuery_CastRayAll(query,
+    &origin, &direction,
+    collector,
+    NULL,
+    NULL,
+    NULL);
+  size_t hit_count;
+  JPH_RayCastResult* hit_array = JPH_AllHit_CastRayCollector_GetHits(collector, &hit_count);
+  for (int i = 0; i < hit_count; i++) {
+    float x = x1 + hit_array[i].fraction * (x2 - x1);
+    float y = y1 + hit_array[i].fraction * (y2 - y1);
+    float z = z1 + hit_array[i].fraction * (z2 - z1);
+    // todo: assuming one shape per collider; doesn't support compound shape
+    Collider* collider = (Collider*) JPH_BodyInterface_GetUserData(
+      world->body_interface,
+      hit_array[i].bodyID);
+    size_t count;
+    Shape** shape = lovrColliderGetShapes(collider, &count);
+    const JPH_RVec3 position = {.x = x, .y = y, .z = z};
+    JPH_Vec3 normal;
+    JPH_Body_GetWorldSpaceSurfaceNormal(collider->body, hit_array[i].subShapeID2, &position, &normal);
+    bool shouldStop = callback(
+      shape[0], // assumes one shape per collider; todo: compound shapes
+      x, y, z,
+      normal.x, normal.y, normal.z,
+      userdata);
+    if (shouldStop)
+      break;
+  }
+  JPH_AllHit_CastRayCollector_Destroy(collector);
+}
+
+static bool lovrWorldQueryShape(World* world, JPH_Shape* shape, float position[3], float scale[3], QueryCallback callback, void* userdata) {
+  JPH_RMatrix4x4 transform;
+  JPH_Vec3 direction = { 0.f };
+  JPH_RVec3 base_offset = { 0.f };
+  const JPH_NarrowPhaseQuery* query = JPC_PhysicsSystem_GetNarrowPhaseQueryNoLock(world->physics_system);
+  float transformArray[16] = {
+    scale[0], 0.f, 0.f, 0.f,
+    0.f, scale[1], 0.f, 0.f,
+    0.f, 0.f, scale[2], 0.f,
+    position[0], position[1], position[2], 1.f
+  };
+  array_to_rmatrix_struct(transformArray, &transform);
+  JPH_AllHit_CastShapeCollector_Reset(cast_shape_collector);
+  JPH_NarrowPhaseQuery_CastShape(query, queryBox, &transform, &direction, &base_offset, cast_shape_collector);
```

Comment:
I'm curious: would it be better to use `CollideShape` for these shape queries, since the direction is unused?

Reply:
In future I would like to expand the current API to support casting, so that shape query is just a special case of shape casting with zero direction.